charms/worker/k8s/src/charm.py

```diff
@@ -577,51 +595,76 @@ def _reconcile(self, event: ops.EventBase):
             self._copy_internal_kubeconfig()
             self._expose_ports()
 
-    @on_error(
-        ops.WaitingStatus("Cluster not yet ready"),
-        AssertionError,
-        subprocess.CalledProcessError,
-        InvalidResponseError,
-        K8sdConnectionError,
-    )
     def _update_status(self):
         """Check k8s snap status."""
-        if self.is_dying:
-            status.add(ops.WaitingStatus("Preparing to leave cluster"))
-            return
-        if self.is_control_plane:
-            assert self.api_manager.is_cluster_ready(), "control-plane not yet ready"  # nosec
-
-        if version := self._get_snap_version():
+        if version := snap_version("k8s"):
             self.unit.set_workload_version(version)
 
+        if not self.get_cluster_name():
+            status.add(ops.WaitingStatus("Node not Clustered"))
+            return
+
         if not self._is_node_ready():
-            status.add(ops.WaitingStatus("Node not yet Ready"))
+            status.add(ops.WaitingStatus("Node not Ready"))
             return
 
-    def _evaluate_removal(self, event: ops.EventBase):
+    def _evaluate_removal(self, event: ops.EventBase) -> bool:
         """Determine if my unit is being removed.
 
         Args:
             event: ops.EventBase - event that triggered charm hook
+
+        Returns:
+            True if being removed, otherwise False
         """
-        if self.is_dying:
-            return
-        if unit := _cluster_departing_unit(event):
+        if self._stored.is_dying:
+            pass
+        elif (unit := _cluster_departing_unit(event)) and unit == self.unit:
             # Juju says I am being removed
-            self._stored.removing = unit == self.unit
```

Comment:
this adjustment let's us keep evaluating the `elif` chain on relation-departed hooks.   This comes in handy when workers handled a relation-departed hook because the last control-plane unit is departing the relation.  the departing unit is the `k8s/N` -- but since its the last one, the k8s-worker needs to die as well